File: java/src/Bj2878.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Bj2878 {
    static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long REMAIN = (long) Math.pow(2, 64);
        int m = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());
        long[] friends = new long[n];

        long sum = 0;
        for (int i = 0; i < n; i++) {
            long num = Long.parseLong(br.readLine());
            friends[i] = num;
            sum += num;
        }

        long remain = sum - m;
        Arrays.sort(friends);

        long angry = 0;
        for (int i = 0; i < n; i++) {
            long removeCount = Math.min(friends[i], remain / (n - i));
            remain -= removeCount;
            friends[i] = removeCount;
            angry += removeCount * removeCount;
            angry %= REMAIN;
        }

        System.out.println(angry);
    }
}

File: kotlin/src/Bj2878.kt
import java.io.BufferedReader
import java.io.InputStreamReader

class Bj2878 {
    companion object{
        fun solve(){
            val REMAIN = Math.pow(2.0, 64.0).toLong()
            val br = BufferedReader(InputStreamReader(System.`in`))
            val (m, n) = br.readLine().split(" ").map { it.toInt() }
            val friends = Array(n) { 0L }
            for (i in 0 until n) {
                friends[i] = br.readLine().toLong()
            }

            var remain = friends.sum() - m
            friends.sort()

            var angry = 0L
            friends.mapIndexed { index, candy ->
                friends[index] = Math.min(candy, remain / (friends.size - index))
                remain -= friends[index]
            }

            friends.map {
                angry += it * it
            }
            angry %= REMAIN
            println(angry)
        }
    }
}